### ISSUE TITLE
Design node state ("memory")

### ISSUE THREAD

**Issue by mvesin** [fedbiomed/fedbiomed#811]
As a developer, I want to design how node state ("memory") should be handled because:

1. it needs to be implemented to protect optimizers' auxiliary variables (node => researcher) with secure aggregation, as we do for model weights (part of #467, dedicated issue to be created)
2. it needs to be clarified for *Design an API for including preprocessing in `Experiment`* #800
3. it may be useful for other purposes in the future

Context:

- the initial model for Fed-BioMed training on node is that each training on a node receives all the necessary inputs from the researcher (model, model parameters, hyper parameters). The node keeps no state ("memory") from execution of previous rounds. The node is not even aware if a training is related or not to a previous training it executed.
- a limited and dedicated type of node state was already implemented as a requirement for secure aggregation. A secure aggregation context (server key or biprime) keeps track of coordinated crypto material between the parties of an experiment.

**Comment by mvesin** [comment]
First set of specifications for implementing initial version:

A node state is a set of data about the status of an experiment *after* a payload (~training/validation for now) execution on a node 
    - key:
        - use state_id (not round number), state_id unique for a *payload* (eg: replaying a round after a breakpoint is a different payload)
        - state_id generated by node and returned to researcher (one per node when executing a payload)
        - None is used for first round, new node
    - content:
        - purpose: a node state is for saving the information needed by the "next round" (next payload ?) to execute (or a payload requested from breakpoint load ?)
        - other content: aside from this needed content, a node state may contain optional information (eg: also received from researcher). Goal: double check, informational, provision for future extensions ?
        - type of data: can contain information like: auxiliary variables (probably needed) and full optimizer state, secagg ids (probably optional), dataloader state (#418), torch persistent buffers (#529)
        - for now, just implement auxiliary variables (also save job_id)
    - security: can be used only by payloads with correct job_id
    - versioning for node state
    - implement breakpoint extension for node state
    - choice for DB techno:
        - for now, continue with tinyDB
        - use distinct table in same tinyDB file
        - pay attention to use same number of params for each entry "similar" (provision for migration to relational DB later ?)
        - save model parameters & bigger data as files, not in DB
    - implementation
        - API: consider adapting `fedbiomed.common.secagg_manager.*SecaggManager`
        - implement in `fedbiomed.node`
        - pass argument as hierarchical/extendable, dict of `state > optimizer > optimodules > auxiliary variables` 
        - tinyDB: no key, make sure the state_id is unique in the DB
    - node implementation recommandations for handling of node state:
        - node state may be missing or incomplete on the node => node has to handle
            - 
        - researcher may send some other value for a field of the stat => researcher value overrides node state
    - job_id is the reference to list all the info (state DB, secagg DB) about one experiment (eg: for listing or cleaning)
    - no cleaning implemented in first version
        - should be easy to later add cleaning function by job_id from researcher

**Comment by sharkovsky** [comment]
we discussed the possibility of having an API for researchers that wish to trigger saving some object as part of the node state between rounds. Here are some possible use cases where this would be useful:
- imagine Scaffold was not implemented in Fed-BioMed, and a researcher wants to implement it. The natural thing would be to follow the implementation given in the original paper, where they rely on saving aux variables as part of the node state between rounds
- researcher wants to test a new optimizer with a different state (e.g. some additional momentum-like variables)
- there are several `persistent buffers` in pytorch that are currently not saved between rounds. A prime example is the batch normalization parameters. Indeed, we have issue #529 open to remind us that we should document this. I imagine that some of these persistent buffers will be automatically saved by our implementation, but I don't think we can foresee all possible situations of interest to researchers

For the uses cases above, I would argue that an API allowing the researcher to interact with the node state is desirable. In my head, it would include functions like `TrainingPlan.save_to_node_state(id: str, obj: Serializable) -> None` and `TrainingPlan.load_from_node_state(id: str, round: Optional[int] = None) -> Serializable`.

**Comment by ybouilla** [comment]
Changes introduced in Researcher

1. in `Job`

```


def collect_node_state(self, round_number : int ,Responses):
       # extract couple (node_id, state_id) from  Responses
      self._state_collection = {round_number: { node_id: state_id}}

def get_last_state(self) -> Dict[str, str]:
     # return last sate_id for each node that have lately responded. If a Node havenot participated, send None
     last_state = {}
    for i in range(len(self._state_collection) , 0, -1 ):
         for node_id in  self._state_collection[i]:
                 last_state [ node_id] = self._state_collection[i].get(node_id)
     return last_state
    
def initiate_state_collection(self):
       self._state_collection : Dict[str, Dict[str, str]] = {}
       self._state_collection[0] = {}
       for node_id in self._data:
               self._state_collection[0] [node_id] = None

**Comment by ybouilla** [comment]
![node state drawio](https://github.com/fedbiomed/fedbiomed/assets/114240274/13bc1fb3-c55f-402e-9698-650e32e5a252)

**Comment by mvesin** [comment]
Thanks @ybouilla for the nice sketch of node/researcher interactions !

One point: for node2 in round3, I would strongly suggest using `state_id=None` instead of `state_id=1224`.

Reason: a state makes sense in the context of a specific payload. In this case, `state_id=1224` was created in the context of the execution of round1, we have no guarantee it makes sense for round2 (eg we save some scaffold state in round1 but don't use scaffold anymore in round2, or the opposite).

So this means: treat a node that was not available or not selected in the previous round as a new node.

**Comment by ybouilla** [comment]
Hi Marc, thanks for your answer. Actually, you mentioned a point that may need more discussion ! :) 
Actually, there might be some cases where what you said  make sense, but if we take the case of scaffold, in the paper, authors are selecting a few Nodes among the pool of Node (`S` being the sample of selected Nodes). For scaffold implementation, you don't want to lose the auxiliary variables of each nodes that havenot communicated during the previous round. 

But maybe this detail should belong to the `Strategy` implementation, and that for now we want to keep things simple using `defaultstrategy`

**Comment by ybouilla** [comment]
![Node state proposal drawio](https://github.com/fedbiomed/fedbiomed/assets/114240274/aba0e5fc-6088-4957-876f-c059264f92e2)

**Comment by ybouilla** [comment]
**update 08.29**

We will create methods in Rounds:
`load_round_state`: where we will detail all the element that need to be saved in a Round
and `save_round_state` (self explanatory) 

ex: 

```python
def save_round_state(self):
      states = {}
      optim_state = self.optimizer.get_state()
      Serializer.dump(optim_state, path)
 ```

**Comment by ybouilla** [comment]
**Updates**:

 - **optimizer type** checking  will help make sure user has not changed its optimizer from one round to another
 - we have to find a way to know how user has changed the optimizer parameters from one round to another. For `declearn` optimizers, this can be done using config (static parameters that user may change) and state (optimizer state, should be internal to FBM), which are well separated. For native torch optimizers, process is more complicated since we don't have such distinction

**Comment by ybouilla** [comment]
**additional work**

> imagine Scaffold was not implemented in Fed-BioMed, and a researcher wants to implement it. The natural thing would be to follow the implementation given in the original paper, where they rely on saving aux variables as part of the node state between rounds
researcher wants to test a new optimizer with a different state (e.g. some additional momentum-like variables)
there are several persistent buffers in pytorch that are currently not saved between rounds. A prime example is the batch normalization parameters. Indeed, we have issue
https://github.com/fedbiomed/fedbiomed/issues/529 open to remind us that we should document this. I imagine that some of these persistent buffers will be automatically saved by our implementation, but I don't think we can foresee all possible situations of interest to researchers

We also have to consider the fact that thanks to Node saving state, it is possible to have a true validation dataset, and not a randomly generated one at each Round, so we can check things like loss function over-fitting and so on...
